Rendering web notifications to the UI and Notification Inbox

aquibbaig opened the pull request:
This is subject to further discussions regarding the design in the meeting today. I am adding that in the agenda. 
TODO: Add `Mark as seen` and/or `Remove Notification` buttons

aquibbaig commented:
<img width="1165" alt="Screenshot 2019-07-08 at 3 11 47 PM" src="https://user-images.githubusercontent.com/26324376/60800740-516eef80-a193-11e9-9d41-a4337f1ff835.png">

aquibbaig commented:
Progress: Added `Mark as seen` and `Remove` Button
<img width="1170" alt="Screenshot 2019-07-09 at 10 26 27 PM" src="https://user-images.githubusercontent.com/26324376/60908089-afd1c600-a298-11e9-817c-f66d542b0e0a.png">

aquibbaig commented:
Just had a check on a mobile viewport too

<img width="321" alt="Screenshot 2019-07-09 at 10 29 01 PM" src="https://user-images.githubusercontent.com/26324376/60908227-fb846f80-a298-11e9-965a-9e889869f51a.png">

imphil commented:
Thanks for the UI proposal. I think we can simplify it a bit. How about this:

- Only show unseen notifications in the box.
- Remove the "Delete" functionality from the popup.
- Only keep the "seen" option.
- Remove the "eye" icon on the left, as now only unseen notifications are presented.
- Add time to the date.
- Make date/time less prominent, and use "relative dates" (we already include a library for that)

Furthermore, make sure that the notifications are "clickable", bringing the user to the "notification inbox". Do you have a screenshot for this page already?

aquibbaig commented:
@imphil I had not given much thought to the "Notification inbox" yet and that's the reason why you can see that it is not implemented yet. So, we would want a Notification inbox page where all the Notifications of the User reside along with these 2 handlers to mark them seen and remove them?

aquibbaig commented:
Also, the CSS in the first commit needs to be fixed. I will handle that!

imphil commented:
Sounds about right, but always try to think about simplifying the user experience. For example, you can avoid a "mark as seen" button if you just mark a notification as seen by clicking on it -- that's pretty standard in any email program. And the delete button is rather rarely used, so make it sufficiently small.

aquibbaig commented:
Thanks, @imphil for your suggestions on this pull request. After the meeting, I switched to this UI, let me know if there is anything bugging you in the design.
<img width="1176" alt="Screenshot 2019-07-10 at 11 59 10 PM" src="https://user-images.githubusercontent.com/26324376/60994874-014c8480-a36f-11e9-93ad-d35e7e5318e1.png">

imphil commented:
The UI looks good now, with a minor nit on the text at the bottom: "See all Notifications" -> "show all notifications"

imphil commented:
I'll have a look at the code early tomorrow (UK time)

aquibbaig commented:
@imphil Regarding routing inside the controllers, there are two things to design now
- Notification Inbox
- Notification Settings

These two things are user preferences in my opinion and should be routed to /user/notification/settings and /user/notification/inbox rather than normal ones. So do you prefer moving the routing logic to UserController instead?

aquibbaig commented:
Adding some css fixes right now to separate seen notifications from unseen!

aquibbaig commented:
Can't seem to do that at the moment cause I can't open any page rn due to
`algoliaException\Hosts Unreachable`, which is most likely a result of either bad internet or some issues with my algolia app that I had created. I'll look into it first thing tomorrow morning

imphil commented:
On routing: We cannot use `/notification`, as this is outside of our available namespace for site pages. `/user/notifications` or `/user/inbox` is fine for the inbox, and settings should be `/user/settings/notifications`, in line with all the other settings pages we have today.

aquibbaig commented:
<img width="1170" alt="Screenshot 2019-07-12 at 11 18 57 PM" src="https://user-images.githubusercontent.com/26324376/61148831-4a324380-a4fd-11e9-8ecc-e2f36add0c3e.png">

# Initial Design for the Notification Inbox
- Unseen notifications have a dark bg, also have a mark as seen and a remove handlers
- Seen notifications only have a remove handler and have transparent bg

aquibbaig commented:
@imphil Just moved the Notification Inbox logic to UserController. Will attempt the settings page tomorrow! Can you please have a look at what's been done already here, when you're free?

aquibbaig commented:
SInce the cross icon clashes with the markasseen cross in the dropdown, let me just quicky change it to a trash icon to symbolise remove action

imphil commented:
First UI comments:

![image](https://user-images.githubusercontent.com/1467123/61223204-272dac80-a714-11e9-8141-1f15b26dd25b.png)

aquibbaig commented:
@imphil  Added that details

> First UI comments:
> 
> ![image](https://user-images.githubusercontent.com/1467123/61223204-272dac80-a714-11e9-8141-1f15b26dd25b.png)

imphil commented:
Initial comments:

UI: Still no padding!
![image](https://user-images.githubusercontent.com/1467123/61599009-84e25d00-ac1c-11e9-979d-8743d359f267.png)


General:
- Notifications -> notifications wherever it appears in a sentence (it's a normal word!)


Further testing depends on https://github.com/librecores/librecores-web/issues/392, I'll look into it (how did you test so far?)

aquibbaig commented:
@imphil The padding was there. Just update the branch as such. About `Notifications->notifications` is done now.

aquibbaig commented:
@imphil  For now, the command would create a Notification.
You can test it by running the consumer itself after that.
`php bin/console rabbitmq:multiple-consumer notification` is the command I guess

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+                            <div class="col-sm-8 col-lg-8 col-xs-8 col-sm-8 inbox-message">
+                                {{ notification.notification.message }}
+                            </div>
+                            <button class="col-xs-4 col-md-4 col-sm-4 text-right notification-remove"
```
That column is way too large with 4/12 of the screen. You need a fixed-size column for the remove icon.

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+    </div>
+    <ul class="all-notifications">
+        {# If there are any notifications #}
+        {% if (mgilet_notification_count(app.user)) %}
```
notificationList should be empty in that case, so you don't need to add the special check.

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+        {# If there are any notifications #}
+        {% if (mgilet_notification_count(app.user)) %}
+            {% for notification in notificationList %}
+                {% if (notification.seen) == true %}
```
Don't duplicate the whole block for seen/unseen. Instead, assign a single CSS class depending on the notification.seen property, and use CSS for further styling.

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+                    <li class="inbox-seen">
+                        <div class="row">
+                            <div class="col-sm-8 col-lg-8 col-xs-8 col-sm-8 inbox-message">
+                                {{ notification.notification.message }}
```
You need to add escaping for HTML special characters, and make sure that newlines get converted into HTML newlines.

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+                {% if (notification.seen) == true %}
+                    {# Render Seen Notification #}
+                    <li class="inbox-seen">
+                        <div class="row">
```
Either use a HTML list, or Bootstrap rows, but not both.

imphil commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+
+{% block javascripts %}
+    {{ parent() }}
+    {{ encore_entry_script_tags('notification') }}
```
Only include the JS for this page here (notification_inbox.js). All other JS should be included in the JS file itself (as import).

imphil commented on site/app/Resources/views/notification/notification_list.html.twig:
```diff
@@ -0,0 +1,33 @@
+{# Render Notifications #}
+{% if mgilet_notification_unseen_count(app.user) != 0 %}
+    {% for notification in notificationList %}
+        {% if notification.seen == false %}
```
You should be able to combine the for and if by using `filter` (https://twig.symfony.com/doc/2.x/filters/filter.html).

imphil commented on site/assets/js/notification.js:
```diff
@@ -0,0 +1,60 @@
+            method: 'post',
+            success: function (response) {
+                if (response === 'seen') {
+                    console.log('Marked notification as seen...');
```
Remove such debug log statements. Leave them in for error conditions.

imphil commented on site/assets/js/notification.js:
```diff
@@ -0,0 +1,60 @@
+        })
+    });
+
+    // Remove a Notification on click
```
That code is unused, isn't it?

aquibbaig replied:
Unfortunately, yes. Just removed it

imphil commented on site/assets/scss/app.scss:
```diff
@@ -982,3 +982,65 @@ a.classifiation-tag:hover {
+.notify-drop .drop-content::-webkit-scrollbar-thumb {
+  background-color: #ccc;
+}
+.notification-unseen {
```
For all these .notification-* classes: Use a single class like `notification` on the outer element (describing a single notification), and then only `unseen`, `seen`, etc. on the sub-elements. You can then do `.notification .unseen`, or nested blocks (it's SASS).

imphil commented on site/assets/scss/app.scss:
```diff
@@ -982,3 +982,65 @@ a.classifiation-tag:hover {
+  padding: 10px 0 5px 0;
+  background-color: #eeeeee;
+}
+.notification-seen{
```
Style: space before {

imphil commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,98 @@
+
+    /**
+     * NotificationController constructor.
+     * @param NotificationManager $notificationManager
```
insert an empty line between the brief and `@param`.

imphil commented on site/webpack.config.js:
```diff
@@ -64,6 +64,12 @@ Encore
     .addEntry('planet', [
         './assets/js/planet.js'
     ])
+    .addEntry('notification', [
```
Try to be consistent: the template is named "notification_list", how about using the same name here?

imphil commented on site/assets/js/notification.js:
```diff
@@ -0,0 +1,60 @@
+            success: function (response) {
+                if (response === 'seen') {
+                    console.log('Marked notification as seen...');
+                    window.location.reload();
```
*Either* you do client-side requests (then you need to hide/delete the items on the client side in addition to sending a request to the server), *or* you do normal form submissions (which reload the page, and then the list is freshly rendered with the remaining notifications). But don't do client-side JS and then reload the page.

aquibbaig replied:
I did a client side request and hid the notification. However any insight on how to update the notification count on the side of the notification bell icon (which gets updated on reload)

aquibbaig replied:
- Also, for form submits, we need to send a response to redirect the user to the page he was in currently when he clicks on the button, which we can't know as the dropdown is present in all pages. 
- Client-side request will not change the notification count immediately for which the user needs to refresh the page

aquibbaig replied:
@imphil I fixed the notification dropdown part and have pushed the changes.
I think since notification inbox is a separate page we can use form submit here

imphil commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,98 @@
+     */
+    public function removeAction(Request $request)
+    {
+        if ($request->isXmlHttpRequest()) {
```
Return early, don't have long-ranging if statements.

```
if (!$request->isXmlHttpRequest()) {
  return new JsonResponse('Not an ajax request');
}
... continue here
```

imphil commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,98 @@
+            $this->notificationManager->removeNotification(
+                [$this->notificationManager->getNotifiableInterface($this->notificationManager->getNotifiableEntityById($notifiable))],
+                $this->notificationManager->getNotification($notification),
+            );
```
Is this needed, isn't the deleteNotification() call below sufficient? If so, then the database schema is missing a couple foreign constraints.

aquibbaig replied:
No, it doesn't. removed it

aquibbaig replied:
Well, yes!
We need to add this and the foreign constraints are provided by the bundle, so no need to worry about that too!

imphil commented:
![image](https://user-images.githubusercontent.com/1467123/61656379-8f523480-acb8-11e9-9c68-2b173624cb33.png)

aquibbaig commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,77 @@
+{% block content %}
+    <h1>Notification Inbox</h1>
+    <div class="inbox-header row">
+        <div class="col-xs-6 col-md-6 col-sm-6 inbox-title">
```
Yes, I did

aquibbaig commented:
@imphil I have switched to form actions in case of the notification inbox whereas ajax works perfectly fine for the notification-list. I have also modified the NotificationController to handle those request.

imphil commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,131 @@
+    {
+        $this->notificationManager->markAsSeen(
+            $this->notificationManager->getNotifiableInterface($this->notificationManager->getNotifiableEntityById($notifiable)),
+            $this->notificationManager->getNotification($notification),
```
You can store the two parameters to the markAsSeen function in a variable to keep the line shorter and more readable. And re-use this variable later when getting the list.

aquibbaig commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * Mark a Notification as seen
+     *
+     * @Route("/notification/seen", name="notification_mark_seen")
```
Will Notification Inbox be a part of User Controller?
SInce routes will now be changed to /user/notification
If that is the case, are we going to move the logic to userController?

imphil replied:
That would make sense, yes.

aquibbaig commented:
@imphil Actually, clicking on a notification means that it has some content that the user will be redirected to. For example, in Facebook it implies a post, In gmail it implies entire messages. So, I guess putting a mark as seen and remove icons to do the same is fine. What do you think?

aquibbaig commented:
@imphil The ajax requests in the notification list return the count variable to the template. So, we handle the entire component after the ajax request has finished successfully. However in the notification inbox, we use forms to actually refresh the page which reflects changes in the notification list above and the list of notifications. It redirects to the notificatio inbox after the request. If we use the same routes(as in ajax request), we will have to handle two components(list and inbox) with js.

imphil commented:
>  So, I guess putting a mark as seen and remove icons to do the same is fine.

I'm not a huge fan of having two buttons there. Maybe another option would be to remove the "delete" button and only leave the mark as seen button?

aquibbaig commented on site/app/Resources/views/base.html.twig:
```diff
@@ -54,6 +54,29 @@
+            <ul class="notification-menu dropdown-menu notify-drop">
+              <div class="notify-drop-title">
+                <div class="row">
+                  <div class="col-md-6 col-sm-6 col-xs-6 notification-title">notifications</div>
```
@imphil I have no idea why GitHub is showing this. But this is already fixed with a rebase

aquibbaig replied:
I'm fixing this in the next commit

aquibbaig commented:
<img width="1188" alt="Screenshot 2019-07-30 at 3 38 20 PM" src="https://user-images.githubusercontent.com/26324376/62121082-87117f00-b2e0-11e9-92ac-b1d19a786de1.png">

@imphil Just tested with long texts such as this. Also now clicking on a notification marks it seen

aquibbaig commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,62 @@
+                                    data-notifiable="{{ notification.notifiableEntity.id }}">
+                                <i class="fa fa-fw fa-times markasseen-icon"></i>
+                            </button>
+                        </form>
```
@imphil This form just contains one button, Do we still need to make it using form builder?

aquibbaig commented on site/src/Controller/NotificationController.php:
```diff
@@ -0,0 +1,131 @@
+    public function markAsSeenAction(Request $request)
+    {
+        $notification = $request->get('notification');
+        $notifiable = $request->get('notifiable');
```
@imphil Now the logic is moved to userController where we fetch the user in the controller itself rather than sending it with the request. We get the notifiable accordingly there. Please have a check is the current approach alright. I have tested and it seems to work well

aquibbaig commented on site/app/Resources/views/notification/notification_inbox.html.twig:
```diff
@@ -0,0 +1,62 @@
+                    </li>
+                    <li class="inbox-message">
+                        {{ notification.notification.message|nl2br }}
+                        {# Mark as seen button #}
```
Now, the user can click on a notification to mark it seen

agathver commented:
IMO, it would be to leave out the border around the buttons. The buttons to clear, or mark as read are not primary actions and hence do not need to be highlighted as such

imphil commented on site/app/Resources/views/user/notification_inbox.html.twig:
```diff
@@ -0,0 +1,54 @@
+        </div>
+    </div>
+    {# Render all notifications #}
+    <div class="all-notifications list-group">
```
- Remove the list-group here (use it on ul only)
- Add a "row" attribute to have the same padding as the previous div.

imphil commented on site/assets/scss/app.scss:
```diff
@@ -982,3 +982,62 @@ a.classifiation-tag:hover {
+  background-color: transparent;
+  border: 0;
+}
+.notification .markasseenIcon:hover {
```
We don't use lowerCamelCase in CSS classes, we use dashes. So this should be .markasseen-icon

imphil commented on site/assets/scss/notification_inbox.scss:
```diff
@@ -0,0 +1,49 @@
+span.inbox-subject, p.inbox-message, p.inbox-notification-date {
+  padding-left: 5px;
+}
+ul.list-group {
```
Don't change the behavior for all ul.list-group elements.

imphil commented on site/src/Controller/UserController.php:
```diff
@@ -223,4 +228,143 @@ public function resendConfirmationEmailAction(
+     *
+     * @param NotificationManager $notificationManager
+     *
+     * @return string;
```
remove the ;

imphil commented on site/src/Controller/UserController.php:
```diff
@@ -223,4 +228,143 @@ public function resendConfirmationEmailAction(
+        $notifiable = $notificationManager->getNotifiableInterface($notificationManager->getNotifiableEntityById($notifiableEntity));
+
+
+        // Deletes the link between notification and notifiable
```
I think I asked this before: Why don't take foreign constraints care of deleting all parts of a notification? If they do (as they should), you should be OK with only one remove* function call. Having only one such call is really helpful: now your remove becomes a single transaction and the database takes care that no intermediate state becomes visible (e.g. a notification being half-deleted).

aquibbaig replied:
@imphil We will resolve this discuss in the meeting today. It's a bit tricky one!

aquibbaig replied:
Added a comment in the code as discussed

aquibbaig commented:
@imphil just edited the commits w.r.t. yesterday's discussion. Let me know what you think

imphil commented on site/app/Resources/views/notification/notification_list.html.twig:
```diff
@@ -0,0 +1,32 @@
+  {% for notification in notificationList|filter(notification => notification.seen == false) -%}
+    {#Render a unseen notification #}
+    {% if (notification.seen == false) %}
+      <ul class="unseen-notifications" id="notification-{{ notification.notification.id }}">
```
Please don't misuse ul/li as generic block element here and in the base.html.twig. Use lists when there are actually things to list (like a list of notifications), but use <div> if you only have blocks next to each other.

aquibbaig replied:
@imphil I had some checks on how others do it and this seems to be a general case(for bootstrap > 3.0). We just have a twig renderer that contains <li> items. <li> can have a nested <ul> and then <li> inside as per documentation. Please have a small look at this [Bootstrap snippet](https://bootsnipp.com/snippets/1e8Ar). So, are we misusing it still?

imphil replied:
ul/li is semantic markup for a list. A list of notifications is a list. A list of elements in a navigation bar is a list. Use ul/li there. But the code you have here is not a list: it's just different paragraphs: a subject, a message, a date, a divider. That's not a list, and therefore it should not use ul/li.

aquibbaig replied:
@imphil Sounds logical indeed, I'm using `<span>` and `<div>`s

imphil commented on site/app/Resources/views/user/notification_inbox.html.twig:
```diff
@@ -0,0 +1,54 @@
+{% extends 'layout.html.twig' %}
+
+{% block title %}
+    Notification Inbox
```
Please indent this file by two spaces and double-check all files this time please.

aquibbaig replied:
Sure!

imphil commented on site/src/Controller/UserController.php:
```diff
@@ -223,4 +228,146 @@ public function resendConfirmationEmailAction(
+
+        // Remove the Notification from the database
+        // TODO: Let foreign constraints take care of removing notifications, see #400
+        // $notificationManager->deleteNotification($notificationManager->getNotification($notification), true);
```
Wait: you commented this out, so it works without this line?

aquibbaig replied:
Yup, it does. As I said yesterday removeNotification() removes link between notifiableEntity and notifiableNotification, so if that is removed the user can't see it anymore

aquibbaig replied:
@imphil The deleteNotification() method is for removing the notification from the database itself

aquibbaig replied:
It is a workaround probably

imphil replied:
So the notification is still half-present in the database? That's not what we want, the notification should be fully gone. Sothe deleteNotification() part is necessary.

aquibbaig replied:
@imphil Exactly!

imphil commented:
Thanks @aquibbaig, I've now merged your PR!

aquibbaig commented:
Final Preview
<img width="1226" alt="Screenshot 2019-08-20 at 10 12 24 PM" src="https://user-images.githubusercontent.com/26324376/63366584-b70be980-c397-11e9-80ec-0325f6f583d9.png">
